src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.cs

```diff
@@ -336,5 +335,71 @@ private void ReturnRangeList(List<(int offset, int length)> ranges)
         {
             _rangeListAllocation = ranges;
         }
+
+        private CborReaderCheckpoint CreateCheckpoint()
+        {
+            return new CborReaderCheckpoint(
+                buffer: _buffer,
+                bytesRead: _bytesRead,
+                depth: _nestedDataItems?.Count ?? 0,
+                contextOffset: (_nestedDataItems == null || _nestedDataItems.Count == 0) ? 0 : _nestedDataItems.Peek().bytesRead,
+                remainingDataItems: _remainingDataItems,
+                isEvenNumberOfDataItemsRead: _isEvenNumberOfDataItemsRead,
+                isTagContext: _isTagContext);
+        }
+
+        private void RestoreCheckpoint(CborReaderCheckpoint checkpoint)
+        {
+            if (_nestedDataItems != null)
+            {
+                int stackOffset = _nestedDataItems.Count - checkpoint.Depth;
+
+                Debug.Assert(
```

Comment:
To consider: should we mark the checkpoint methods public or protected? If so these asserts should be changed to runtime checks.

Reply:
Making it public would require further validation, like that the checkpoint came from this instance (otherwise the stack wouldn't have the same meaning).

You'd also need to account for more complicated things, like the stack being popped and a new frame pushed (complete the current array, start the next array/map/tag, then rewind) which there's currently not a need for (and what happens if you restore a checkpoint to a different reader instance?).

So, I'd say "if a scenario warrants it, OK, but I'd hold off unless it's clearly needed; then you need to decide if you're using the public checkpoint internally, or this sort of slim checkpoint")

Reply:
> You'd also need to account for more complicated things, like the stack being popped and a new frame pushed.

That's already accounted for by storing the buffer offset in each frame. It's currently only being used in the assert.

> So, I'd say "if a scenario warrants it, OK, but I'd hold off unless it's clearly needed; then you need to decide if you're using the public checkpoint internally, or this sort of slim checkpoint")

I feel this is already in the borderline of over-engineering, but I thought I'd get a second opinion regardless :-)